draft-thomson-http-oblivious.md

```diff
@@ -159,10 +159,110 @@ occur, as shown in {{fig-overview}}:
+An encapsulated HTTP message includes the following values:
+
+1. A binary-encoded HTTP message [CITEME].
+2. Padding of arbitrary length which MUST contain all zeroes.
+
+The encoding of an HTTP message is as follows:
+
+~~~
+Plaintext Message {
+  Message Length (i),
+  Message (..),
+  Padding Length (i),
+  Padding (..),
+}
+~~~
+
+This structure is then encrypted under a key with a specific identity, forming an encapsulated HTTP
+message, with the following structure:
+
+~~~
+Key Identifer {
+  Key ID Length (i),
+  Key ID (..),
+}
+
+Encapsulated Message {
+  Key Identifier (..)
+  Encrypted Message Length (i),
+  Encrypted Message (..),
+}
+~~~
+
+The encapsulated message Key ID, as well as the encryption mechanics, are different for requests
+and responses, as described below.
+
 ## HPKE Encapsulation of Requests {#request}
 
+Clients encapsulate a request Plaintext Message `msg` with an HPKE public key `pkR`, whose Key Identifier
+is `keyID` as follows:
+
+1. Compute an HPKE context using `pkR`, yielding `context` and encapsulation key `enc`
+2. Encrypt (seal) `msg` with `keyID` as associated data using `context`, yielding ciphertext `ct`
+3. Concatenate `enc` and `ct`, yielding an Encrypted Message `encrypted_msg`
+
+In pseudocode, this procedure is as follows:
+
+~~~
+enc, context = SetupBaseS(pkR, "request")
+aad = 0x01 || keyID
+encrypted_msg = context.Seal(aad, msg)
+~~~
+
+Clients construct the Encapsulated Message `req` using `keyID` and `encrypted_msg`.
+
+Servers decrypt an Encapsulated Message by reversing this process. Given an Encapsulated Message `req` request
+with Key Identifier `keyID` corresponding to an existing HPKE private key `skR`, servers decapsulate
+the Message as follows:
+
+1. Parse the `req` Encrypted Message as the concatenation of `enc` and `encrypted_message`
+2. Compute an HPKE context using `skR` and the encapsulated key `enc` from `req`, yielding `context`
+3. Decrypt `encrypted_message` with `keyID` as associated data, yielding `msg` or an error on failure
+
+In pseudocode, this procedure is as follows:
+
+~~~
+context = SetupBaseR(enc, skR, "request")
+aad = 0x01 || keyID
+msg, error = context.Open(aad, ct)
+~~~
+
+Servers MUST verify that the Plaintext Message padding consists of all zeroes before processing the
+corresponding Message.
+
 ## HPKE Encapsulation of Responses {#response}
 
+Given an HPKE context `context` and a response Plaintext Message `resp` sent in response to a Plaintext
+Message `req`, servers encrypt the data as follows:
+
+1. Derive a symmetric key and nonce from `context`
+2. Encrypt `resp` with empty Key Identifier `emptyKeyID` as associated data, yielding `encrypted_msg`
+
+In pseudocode, this procedure is as follows:
+
+~~~
+secret = context.Export("secret", 32)
+prk = Extract(req, secret)
+key = Expand(secret, "key", Nk)
+nonce = Expand(secret, "nonce", Nn)
+aad = 0x02 || emptyKeyID
+encrypted_msg = Seal(key, nonce, aad, resp)
```

Comment:
```suggestion
key = context.Export("response key", Nk)
nonce = context.Export("response nonce", Nn)
aad = ???
encrypted_msg = Seal(key, nonce, aad, resp)
```

My preference would be to export two values for this.

Reply:
We considered this for ODoH too, but this removes binding between the response encryption key material and the request. So, in theory, it would be possible for an application to mistakenly send the same `enc` to the server, forcing it to derive the same key and nonce, and possibly re-use them for encrypting the response. I think we should bake key/nonce re-use into the encapsulation mechanism, if possible.

Reply:
Oh, I missed the fact that you are including the entire request (?) in the key schedule.  That seems like something that is worth documenting.  I missed it completely.

However, it might be enough to include `enc`, so that you don't have to remember the entire message.

Reply:
Even `enc` might not be enough, since a client could mistakenly re-use `enc` for more than one request, leading to the nonce/key reuse issue, I think.

Reply:
At that point you have nonce reuse for the request, so any protection you have for the response is far less valuable.

Reply:
Hmm, true, though only if the request contents are different, right? Either way, if enc is reused across requests, it seems:

1. If the request is different, then we have reuse on the request
2. If the request is the same, and the response is different, then we have reuse on the response. 
3. If the request and responses are the same, then no reuse

Unless I'm missing something, (1) seems something we may have to just acknowledge. I don't see a way to prevent it. One might try to include a nonce in the `context` derivation, but if the client reuses `enc`, then it might just reuse that nonce. (ODoH avoided this since the requests always had some entropy in them, and should therefore be different with high probability.)

So if we want to focus on (2) and (3), where the requests are the same, then it seems (2) should be the assumed case. (Relying on content to never change seems silly.) And the only way to seemingly prevent that is for the server to contribute randomness to the key derivation. 

Given that, I see a number of options:

1. Include request in the key derivation: this doesn't address case (2) above.
2. Include server nonce in the key derivation: this addresses (2), but makes the server entirely responsible for preventing nonce reuse. (Do we trust their randomness?)
3. Include both request and a server nonce in key derivation: this addresses (2), and hedges against bad randomness

What do you think?

Reply:
I think that there is no really good answer here.  Given 0-RTT, the risk of request replay seems like we might want to have some sort of nonce.  Having a nonce in both places makes some sense.  Let's open an issue on this rather than continue here though.